### PR TITLE
{AKS} add `--network-policy` to `az aks update`

### PR DESCRIPTION
**Pull request by robogatikov**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command
`az aks update --network-policy=none`
`az aks update --network-policy=azure`
`az aks update --network-policy=cilium`

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
AKS

File: src/aks-preview/azext_aks_preview/_help.py [hunk 1]
```diff
@@ -854,6 +854,12 @@
           long-summary: |
               Used to control the mode the network plugin should operate in. For example, "overlay" used with
               --network-plugin=azure will use an overlay network (non-VNET IPs) for pods in the cluster.
+        - name: --network-policy
+          type: string
+          short-summary: Update the mode of a network policy.
+          long-summary: |
+              Specify "azure" for Azure network policy manager, "cilium" for Azure CNI Overlay powered by Cilium, and "none" to uninstall "azure" network policy.
+              Defaults to "" (network policy disabled).
```
Comment (wedaly) [inline src/aks-preview/azext_aks_preview/_help.py:862]: I'd suggest removing references to "none" for now, since it's not fully implemented in AKS yet.

File: src/aks-preview/azext_aks_preview/_params.py [hunk 2]
```diff
@@ -466,6 +471,7 @@ def load_arguments(self, _):
         c.argument('nat_gateway_managed_outbound_ip_count', type=int, validator=validate_nat_gateway_managed_outbound_ip_count)
         c.argument('nat_gateway_idle_timeout', type=int, validator=validate_nat_gateway_idle_timeout)
         c.argument('network_dataplane', arg_type=get_enum_type(network_dataplanes))
+        c.argument('network_policy', arg_type=get_enum_type(network_policies))
```
Comment (wedaly) [inline src/aks-preview/azext_aks_preview/_params.py:474]: adding enum validation is nice, but could we defer that to a later PR? Want to avoid referencing "none" until it's ready for people to try it.

File: src/aks-preview/azext_aks_preview/managed_cluster_decorator.py [hunk 3]
```diff
@@ -3151,6 +3158,11 @@ def update_network_plugin_settings(self, mc: ManagedCluster) -> ManagedCluster:
         pod_cidr = self.context.get_pod_cidr()
         if pod_cidr:
             mc.network_profile.pod_cidr = pod_cidr
+
+        network_policy = self.context.get_network_policy()
+        if network_policy:
+            mc.network_profile.network_policy = network_policy
```
Comment (wedaly) [inline src/aks-preview/azext_aks_preview/managed_cluster_decorator.py:3164]: in addition to the live test you added, should we have a unit test in test_managed_cluster_decorator.py? Maybe add something to `test_update_network_plugin_settings`?

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py [hunk 4]
```diff
@@ -4468,7 +4469,7 @@ def test_aks_migrate_cluster_to_cilium_dataplane(self, resource_group, resource_
         ])
 
         # update to enable cilium dataplane
-        update_cmd = 'aks update -g {resource_group} -n {name} --network-dataplane=cilium'
+        update_cmd = 'aks update -g {resource_group} -n {name} --network-dataplane=cilium --network-policy=cilium'
```
Comment (wedaly) [inline src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:4473]: do we need to update the test fixtures after running the live test? I would have expected this to change the PUT MC request, but don't see any changes in `recordings/test_aks_migrate_cluster_to_cilium_dataplane.yaml`

File: src/aks-preview/azext_aks_preview/_help.py (hunk 1 above)
Comment (robogatikov) [reply src/aks-preview/azext_aks_preview/_help.py:862]: Removed mentioning "none"

File: src/aks-preview/azext_aks_preview/_params.py (hunk 2 above)
Comment (robogatikov) [reply src/aks-preview/azext_aks_preview/_params.py:474]: Removed enum validation

File: src/aks-preview/azext_aks_preview/managed_cluster_decorator.py (hunk 3 above)
Comment (robogatikov) [reply src/aks-preview/azext_aks_preview/managed_cluster_decorator.py:3164]: Added a test case to update network-policy from empty string to "azure"

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py (hunk 4 above)
Comment (robogatikov) [reply src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:4473]: I am not sure about test fixtures, probably @FumingZhang or @zhoxing-ms can help clarify

Comment (FumingZhang) [reply src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:4473]: Queued [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=80318220) to validate the change. If test passed, the recording file would be published as pipeline artifact, could download and commit it.

File: src/aks-preview/azext_aks_preview/_help.py [hunk 5]
```diff
@@ -854,6 +854,12 @@
           long-summary: |
               Used to control the mode the network plugin should operate in. For example, "overlay" used with
               --network-plugin=azure will use an overlay network (non-VNET IPs) for pods in the cluster.
+        - name: --network-policy
+          type: string
+          short-summary: Update the mode of a network policy.
+          long-summary: |
+              Specify "azure" for Azure network policy manager, "cilium" for Azure CNI Overlay powered by Cilium.
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/_help.py:861]: So there's no way to update the network policy of an existing cluster to calico? And if it's possible to update the network policy to azure, could you please help add a new live test case?

Comment (FumingZhang) [reply src/aks-preview/azext_aks_preview/_help.py:861]: If my understanding is correct, in update command, when user specify option value as `""`, the network policy would be disabled (and in rp response this is `none`), please also add a live test case for this scenario.

File: src/aks-preview/azext_aks_preview/managed_cluster_decorator.py (hunk 3 above)
Comment (FumingZhang) [reply src/aks-preview/azext_aks_preview/managed_cluster_decorator.py:3164]: Python implicitly converts `""` to `False`, if user specifies `--network-policy ""`, the code will not update the property value to `""`, is this expected?

File: src/aks-preview/azext_aks_preview/_help.py (hunk 5 above)
Comment (wedaly) [reply src/aks-preview/azext_aks_preview/_help.py:861]: unfortunately, sending RP networkPolicy="" gets ignored in the update due to partial PUT semantics (AKS RP backfills with the current value in the cluster). 20230802preview API adds a "none" value so cx can explicitly disable the network policy.

Comment (wedaly) [reply src/aks-preview/azext_aks_preview/_help.py:861]: and yes, we will add a live test once we fully support setting networkPolicy="none". RP changes aren't ready yet though.

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py (hunk 4 above)
Comment (FumingZhang) [reply src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:4473]: The test failed due to 
> E           azure.core.exceptions.HttpResponseError: (PropertyChangeNotAllowed) Changing property 'networkProfile.networkPolicy' is not allowed.
> E           Code: PropertyChangeNotAllowed
> E           Message: Changing property 'networkProfile.networkPolicy' is not allowed.
> E           Target: networkProfile.networkPolicy

**Comment by FumingZhang** [comment]
Requeued [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=81039707&view=results)

**Review by FumingZhang — APPROVED** [review]
LGTM

[live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=81039707&view=results) passed

**Review by wedaly — APPROVED** [review]

**Comment by zhoxing-ms** [comment]
Please take a look at this comment https://github.com/Azure/azure-cli-extensions/pull/6809#issuecomment-1736199904

**Comment by robogatikov** [comment]
> Please take a look at this comment [#6809 (comment)](https://github.com/Azure/azure-cli-extensions/pull/6809#issuecomment-1736199904)

Thanks, @zhoxing-ms , added description to HISTORY.rst and bumped extension version to 1.5.161

**Comment by robogatikov** [comment]
> Please fix conflicting files

Thanks @yonzhan , forgot to sync my fork. Resolved merge conflicts

**Review by zhoxing-ms — APPROVED** [review]
